sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py

```diff
@@ -60,7 +64,7 @@ class AsyncHTTPPolicy(abc.ABC, Generic[HTTPRequestType, AsyncHTTPResponseType]):
     """
     def __init__(self) -> None:
         # next will be set once in the pipeline
-        self.next = None # type: Optional[Any]
+        self.next = None # type: Optional[Union[HTTPPolicy, HttpTransport, AsyncHTTPPolicy, AsyncHttpTransport]]
```

Comment:
Sure, some AsyncHTTPPolicy implementations happen to subclass HTTPPolicy (seems to me that's a problem, if perhaps a separate one). But this implies `next()` may not return a coroutine. In addition to that being misleading, I expect mypy to consider `await self.next()` an error given this type.

Reply:
I don't think we require next to be an AsyncHTTPPolicy. 

It is valid that we have a aync_retry_policy w/ a header_policy/user_agent_policy.

Which means, yes, next.send() may not return a coroutine.

Reply:
HeaderPolicy and UserAgentPolicy are both SansIOHTTPPolicy. AsyncPipeline [wraps them in an async runner](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/core/azure-core/azure/core/pipeline/_base_async.py#L147). Do we support or expect HTTPPolicy or HttpTransport in AsyncPipeline?

Reply:
I would focus the PR on what it's trying to solve: the retry policy problem, and not this one. I'm not sure I'm ready to be opiniated here as a side effect of a "retry policy" fix. As @chlowell implies, the choice here is more subtle that it seems and it would requires its own PR (I know, I'm freak about atomic PRs, but I assume it :p)

Reply:
The problem here is we have 

```py
self.next = None # type: Union[HTTPPolicy, HttpTransport]
```

In __HTTPPolicy__

but 

```py
self.next = None # type: Union[AsyncHTTPPolicy, AsyncHttpTransport]
```

in __AsyncHTTPPolicy__

If we make AsyncHTTPPolicy a subclass of HTTPPolicy, mypy is not happy.

@lmazuel

Reply:
Well, what about you leave `Any` for this PR and we talk about that outside of this PR scope? :)

Reply:
OK. I separate it into two PRs. This one will be focusing on cleaning up the hierarchy of HttpPolicy & AsyncHttpPolicy (the other PR depends on this change).